src/osd/PrimaryLogPG.cc

```diff
@@ -5846,21 +5846,23 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
   auto& op = osd_op.op;
   auto& oi = ctx->new_obs.oi;
   auto& soid = oi.soid;
+  uint64_t size = oi.size;
 
-  if (op.extent.truncate_seq) {
-    dout(0) << "sparse_read does not support truncation sequence " << dendl;
-    return -EINVAL;
-  }
+    // are we beyond truncate_size?
+  if ( (oi.truncate_seq < op.extent.truncate_seq) &&
+       (op.extent.offset + op.extent.length > op.extent.truncate_size) &&
+       (size > op.extent.truncate_size) )
+    size = op.extent.truncate_size;
 
   ++ctx->num_read;
   if (pool.info.is_erasure()) {
     // translate sparse read to a normal one if not supported
     uint64_t offset = op.extent.offset;
     uint64_t length = op.extent.length;
-    if (offset > oi.size) {
+    if (offset > size) {
       length = 0;
-    } else if (offset + length > oi.size) {
-      length = oi.size - offset;
+    } else if (offset + length > size) {
+      length = size - offset;
```

Comment:
Should we also do something like this in none erasure case ?

Reply:
Maybe. I'm still seeing some data corruption errors with fsx, and that might be the culprit. I'll play with a revised patch today.

One thing I'm not clear on is that extent.length will end up being reset below around like 5920. Why is that being done?

Reply:
Latest set limits the fiemap call to the truncate_size when truncate_seq is larger than what the MDS has.